lib/ansible/modules/system/known_hosts.py

```diff
@@ -137,25 +142,20 @@ def enforce_state(module, params):
             else:
                 module.fail_json(msg="Failed to read %s: %s" % (path, str(e)))
         try:
-            outf = tempfile.NamedTemporaryFile(mode='w+', dir=os.path.dirname(path))
-            if inf is not None:
-                for line_number, line in enumerate(inf):
-                    if found_line == (line_number + 1) and (replace_or_add or state == 'absent'):
-                        continue  # skip this line to replace its key
-                    outf.write(line)
-                inf.close()
-            if state == 'present':
-                outf.write(key)
-            outf.flush()
-            module.atomic_move(outf.name, path)
+            with tempfile.NamedTemporaryFile(mode='w+', dir=os.path.dirname(path), delete=False) as outf:
+                if inf is not None:
+                    for line_number, line in enumerate(inf):
+                        if found_line == (line_number + 1) and (replace_or_add or state == 'absent'):
+                            continue  # skip this line to replace its key
+                        outf.write(line)
+                    inf.close()
+                if state == 'present':
+                    outf.write(key)
+                outf.flush()
+                module.atomic_move(outf.name, path)
```

Comment:
The call to `atomic_move` should be moved outside the `with` block, using [try...except...else](https://docs.python.org/2/tutorial/errors.html#handling-exceptions) seems appropriate:
```
try:
    with tempfile.NamedTemporaryFile(...) as outf:  # automatically calls "outf.close()"
        [...]
except (IOError, OSError) as e:
    module.fail_json(...)
else:
    # here outf.close() has been called
    module.atomic_move(outf.name, path)  # atomic_move doesn't raise IOError, OSError (they are caught inside the atomic_move method)
```

Reply:
Ok, I've also removed flushing of the outfile since closing the file should take care of that, let's see how the testrun goes.